feat(eslint-plugin): [prefer-readonly-parameter-types] added an optional type allowlist (#4436)
typescript-eslint/typescript-eslint: packages/eslint-plugin/docs/rules/prefer-readonly-parameter-types.md

## Options

```ts
type TypeAllowlistItem = {
  typeName: string;

> } & ({ local: true } | { defaultLib: true } | { package: string });

Reviewer: Putting a little more thought into this: I'm now wondering if it would make more sense to use a traditional [discriminated union](https://basarat.gitbook.io/typescript/type-system/discriminated-unions)...

```ts
} & ({ source: 'default-lib' } | { source: 'local' } | { package: string; source: 'package' });
```

It's a little easier to check in code against the one property. One discriminant type is easier to remember than three boolean/string properties, too...

Thoughts?

cc @bradzacher too as this might end up being a type used across other options.

Author: Yeah, that's quite clearly a better solution.

interface Options {
  allowlist: Array<TypeAllowlistItem>;
  checkParameterProperties?: boolean;
  ignoreInferredTypes?: boolean;
  treatMethodsAsReadonly?: boolean;
}

const defaultOptions: Options = {

>   allowlist: [{typeName: "HTMLElement", defaultLib: true}];

Reviewer: ```suggestion
  allowlist: [];
```

Author: Ah, thanks!

## Options

### `allow`

Some complex types cannot easily be made readonly, for example the `HTMLElement` type or the `JQueryStatic` type from `@types/jquery`. This option allows you to globally disable reporting of such types.

Each item must be one of:

- A type defined in a file (`{from: "file", name: "Foo", path: "src/foo-file.ts"}` with `path` being an optional path relative to the project root directory)

> - A type from the default library (`{from: "lib", name: "Foo"}`)

Reviewer 2: `defaultLib` would be a better `from` value as it would make it more explicit that it refers to the TypeScript default libs rather than just a random lib.

Author: See https://github.com/typescript-eslint/typescript-eslint/issues/5271#issuecomment-1293768915

Reviewer 2: Ahh, that's were all the conversation about the details is. 🙂

## Options

### `allow`

Some complex types cannot easily be made readonly, for example the `HTMLElement` type or the `JQueryStatic` type from `@types/jquery`. This option allows you to globally disable reporting of such types.

Each item must be one of:

> - A type defined in a file (`{from: "file", name: "Foo", source: "src/foo-file.ts"}` with `source` being optional)

Reviewer 3: We should explain what it means to set `source` (i.e. is it relative or absolute?) as well as what it means to not set `source`.

Right now it's ambiguous and unclear what it means, IMO.

Author: Thanks, good catch, better explanation added.

## Options

### `allowlist`

Some complex types cannot easily be made readonly, for example the `HTMLElement` type or the `JQueryStatic` type from `@types/jquery`. This option allows you to globally disable reporting of such types.

> Each item has to be either a type defined in a file (`{from: "file", name: "Foo", source: "src/foo-file.ts"}` with `source` being optional), a type from the default library (`{from: "lib", name: "Foo"}`), or a file from a specific package (`{from: "package", name: "Foo", source: "foo-lib"}`, this also works for types defined in a typings package). Additionally, multiple sources may be combined (`{from: ["file", "package"], name: "Foo"}`) or a type may be defined just as a simple string, which then matches the type independently of its origin.

Reviewer: Putting these points all together in a paragraph hurts readability. I'd recommend switching to a list:

```suggestion
Each item must be one of:
* A type defined in a file (`{from: "file", name: "Foo", source: "src/foo-file.ts"}` with `source` being optional)
* A type from the default library (`{from: "lib", name: "Foo"}`)
* A type from a package (`{from: "package", name: "Foo", source: "foo-lib"}`, this also works for types defined in a typings package).

Two additional formats are supported:
* Multiple sources, as an array for `from` (`{from: ["file", "package"], name: "Foo"}`)
* A type may be defined just as a simple string, which then matches the type independently of its origin.
```

> Examples of code for this rule with `{allowlist: [{source: "file", name: "Foo"}, {source: "lib", name: "HTMLElement"}, {from: "package", name: "Bar", source: "bar-lib"}]}`:

Reviewer: This is a lot of POJO for a single line. How about ...

```suggestion
Examples of code for this rule with:

```json
{
  "allowlist": [
    { "source": "file", "name": "Foo" },
    { "source": "lib", "name": "HTMLElement" },
    { "from": "package", "name": "Bar", "source": "bar-lib" }
  ]
}
```
